File: MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/location/engine/LocationEngineResultTest.java
package com.mapbox.mapboxsdk.location.engine;

import android.content.Intent;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class LocationEngineResultTest {

  private static final double TEST_LAT_LNG = 1.0;

  @Test
  public void checkNullIntent() {
    LocationEngineResult result = LocationEngineResult.extractResult(null);
    assertThat(result).isNull();
  }

  @Test
  public void passInvalidIntent() {
    Intent intent = mock(Intent.class);
    LocationEngineResult result = LocationEngineResult.extractResult(intent);
    assertThat(result).isNull();
  }

  @Test
  public void passValidIntent() {
    Location location = mock(Location.class);
    LocationEngineResult result = LocationEngineResult.extractResult(getValidIntent(location));
    assertThat(result).isNotNull();
    assertThat(result.getLastLocation()).isSameAs(location);
  }

  @Test
  public void passNullLocation() {
    Location location = null;
    LocationEngineResult result = LocationEngineResult.create(location);
    assertForNullInput(result);
  }

  @Test
  public void passNullLocationList() {
    List<Location> locations = null;
    LocationEngineResult result = LocationEngineResult.create(locations);
    assertForNullInput(result);
  }

  @Test
  public void passValidLocation() {
    Location location = getValidLocation();
    LocationEngineResult result = LocationEngineResult.create(location);
    assertForValidInput(result);
  }

  @Test
  public void passValidLocationList() {
    List<Location> locations = Collections.singletonList(getValidLocation());
    LocationEngineResult result = LocationEngineResult.create(locations);
    assertForValidInput(result);
  }

  @Test
  public void passMutableLocationListWithNulls() {
    List<Location> locations = getLocationsWithNulls();
    LocationEngineResult result = LocationEngineResult.create(locations);
    assertForValidInput(result);
  }

  @Test
  public void passImmutableLocationListWithNulls() {
    List<Location> locations = Collections.unmodifiableList(getLocationsWithNulls());
    LocationEngineResult result = LocationEngineResult.create(locations);
    assertForValidInput(result);
  }

  @Test
  public void passImmutableListWithNullLocation() {
    List<Location> locations = Collections.singletonList(null);
    LocationEngineResult result = LocationEngineResult.create(locations);
    assertThat(result != null);
    assertThat(result.getLocations().size() == 0);
  }

  private static List<Location> getLocationsWithNulls() {
    return new ArrayList<Location>() {
      {
        add(null);
        add(getValidLocation());
        add(null);
      }
    };
  }

  private static Location getValidLocation() {
    Location location = mock(Location.class);
    when(location.getLatitude()).thenReturn(TEST_LAT_LNG);
    when(location.getLongitude()).thenReturn(TEST_LAT_LNG);
    return location;
  }

  private static void assertForNullInput(LocationEngineResult result) {
    assertThat(result).isNotNull();
    assertThat(result.getLocations()).isEmpty();
  }

  private static void assertForValidInput(LocationEngineResult result) {
    assertThat(result.getLocations()).isNotNull();
    assertThat(result.getLocations().size()).isEqualTo(1);
    assertThat(result.getLocations().get(0).getLatitude()).isEqualTo(TEST_LAT_LNG);
    assertThat(result.getLocations().get(0).getLongitude()).isEqualTo(TEST_LAT_LNG);
  }

  private static Intent getValidIntent(Location location) {
    Intent intent = mock(Intent.class);
    when(intent.hasExtra(LocationManager.KEY_LOCATION_CHANGED)).thenReturn(true);
    Bundle bundle = mock(Bundle.class);
    when(bundle.getParcelable(LocationManager.KEY_LOCATION_CHANGED)).thenReturn(location);
    when(intent.getExtras()).thenReturn(bundle);
    return intent;
  }
}

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineResultTest.kt
package org.maplibre.android.location.engine

import android.content.Intent
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import org.assertj.core.api.Assertions
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner
import java.util.Collections

@RunWith(MockitoJUnitRunner::class)
class LocationEngineResultTest {
    @Test
    fun checkNullIntent() {
        val result = LocationEngineResult.extractResult(null)
        Assertions.assertThat(result).isNull()
    }

    @Test
    fun passInvalidIntent() {
        val intent = Mockito.mock(Intent::class.java)
        val result = LocationEngineResult.extractResult(intent)
        Assertions.assertThat(result).isNull()
    }

    @Test
    fun passValidIntent() {
        val location = Mockito.mock(Location::class.java)
        val result = LocationEngineResult.extractResult(getValidIntent(location))
        Assertions.assertThat(result).isNotNull()
        Assertions.assertThat(result!!.lastLocation).isSameAs(location)
    }

    @Test
    fun passNullLocation() {
        val location: Location? = null
        val result = LocationEngineResult.create(location)
        assertForNullInput(result)
    }

    @Test
    fun passNullLocationList() {
        val locations: List<Location>? = null
        val result = LocationEngineResult.create(locations)
        assertForNullInput(result)
    }

    @Test
    fun passValidLocation() {
        val location = validLocation
        val result = LocationEngineResult.create(location)
        assertForValidInput(result)
    }

    @Test
    fun passValidLocationList() {
        val locations = listOf(validLocation)
        val result = LocationEngineResult.create(locations)
        assertForValidInput(result)
    }

    @Test
    fun passMutableLocationListWithNulls() {
        val locations = locationsWithNulls
        val result = LocationEngineResult.create(locations)
        assertForValidInput(result)
    }

    @Test
    fun passImmutableLocationListWithNulls() {
        val locations = Collections.unmodifiableList(locationsWithNulls)
        val result = LocationEngineResult.create(locations)
        assertForValidInput(result)
    }

    @Test
    fun passImmutableListWithNullLocation() {
        val locations = listOf<Location?>(null)
        val result = LocationEngineResult.create(locations)
        Assertions.assertThat(result != null)
        Assertions.assertThat(result.locations.size == 0)
    }

    companion object {
        private const val TEST_LAT_LNG = 1.0
        // J2K: IDE suggestion to use ArrayList<Location?>
        private val locationsWithNulls: ArrayList<Location?>
            private get() = object : ArrayList<Location?>() {
                init {
                    add(null)
                    add(validLocation)
                    add(null)
                }
            }
        private val validLocation: Location
            private get() {
                val location = Mockito.mock(Location::class.java)
                Mockito.`when`(location.latitude).thenReturn(TEST_LAT_LNG)
                Mockito.`when`(location.longitude).thenReturn(TEST_LAT_LNG)
                return location
            }

        private fun assertForNullInput(result: LocationEngineResult) {
            Assertions.assertThat(result).isNotNull()
            Assertions.assertThat(result.locations).isEmpty()
        }

        private fun assertForValidInput(result: LocationEngineResult) {
            Assertions.assertThat(result.locations).isNotNull()
            Assertions.assertThat(result.locations.size).isEqualTo(1)
            Assertions.assertThat(result.locations[0].latitude).isEqualTo(TEST_LAT_LNG)
            Assertions.assertThat(result.locations[0].longitude).isEqualTo(TEST_LAT_LNG)
        }

        private fun getValidIntent(location: Location): Intent {
            val intent = Mockito.mock(Intent::class.java)
            Mockito.`when`(intent.hasExtra(LocationManager.KEY_LOCATION_CHANGED)).thenReturn(true)
            val bundle = Mockito.mock(Bundle::class.java)
            Mockito.`when`<Any?>(bundle.getParcelable(LocationManager.KEY_LOCATION_CHANGED)).thenReturn(location)
            Mockito.`when`(intent.extras).thenReturn(bundle)
            return intent
        }
    }
}
